packages/artifact/src/internal-download-http-client.ts

```diff
@@ -12,122 +12,174 @@ import {
-  const client = createHttpClient()
-  let downloadedFiles = 0
-  await Promise.all(
-    parallelDownloads.map(async () => {
-      while (downloadedFiles < downloadItems.length) {
-        const currentFileToDownload = downloadItems[downloadedFiles]
-        downloadedFiles += 1
-        await downloadIndividualFile(
-          client,
-          currentFileToDownload.sourceLocation,
-          currentFileToDownload.targetPath
-        )
-      }
-    })
-  )
-}
+  /**
+   * Fetches a set of container items that describe the contents of an artifact
+   * @param artifactName the name of the artifact
+   * @param containerUrl the artifact container URL for the run
+   */
+  async getContainerItems(
+    artifactName: string,
+    containerUrl: string
+  ): Promise<QueryArtifactResponse> {
+    // the itemPath search parameter controls which containers will be returned
+    const resourceUrl = new URL(containerUrl)
+    resourceUrl.searchParams.append('itemPath', artifactName)
+    this.downloadHttpManager.createClients(1)
+    const client = this.downloadHttpManager.getClient(0)
+    const requestOptions = getRequestOptions('application/json')
+    const rawResponse = await client.get(resourceUrl.toString(), requestOptions)
+    const body: string = await rawResponse.readBody()
+    if (isSuccessStatusCode(rawResponse.message.statusCode) && body) {
+      return JSON.parse(body)
+    }
+    // eslint-disable-next-line no-console
+    console.log(rawResponse)
+    throw new Error(`Unable to get ContainersItems from ${resourceUrl}`)
+  }
 
-/**
- * Downloads an individual file
- * @param client http client that will be used to make the necessary calls
- * @param artifactLocation origin location where a file will be downloaded from
- * @param downloadPath destination location for the file being downloaded
- */
-export async function downloadIndividualFile(
-  client: HttpClient,
-  artifactLocation: string,
-  downloadPath: string
-): Promise<void> {
-  const stream = fs.createWriteStream(downloadPath)
-  const response = await client.get(artifactLocation)
-  if (isSuccessStatusCode(response.message.statusCode)) {
-    await pipeResponseToStream(response, stream)
-  } else if (isRetryableStatusCode(response.message.statusCode)) {
-    warning(
-      `Received http ${response.message.statusCode} during file download, will retry ${artifactLocation} after 10 seconds`
+  /**
+   * Concurrently downloads all the files that are part of an artifact
+   * @param downloadItems information about what items to download and where to save them
+   */
+  async downloadSingleArtifact(downloadItems: DownloadItem[]): Promise<void> {
+    const DOWNLOAD_CONCURRENCY = getDownloadFileConcurrency()
+    // limit the number of files downloaded at a single time
+    const parallelDownloads = [...new Array(DOWNLOAD_CONCURRENCY).keys()]
+    this.downloadHttpManager.createClients(DOWNLOAD_CONCURRENCY)
```

Comment:
If you really want to create a pool of connections, we should:
1. Let the `HttpManager` handle all client creation and disposal.  So as soon as a `httpManager` is initiated, it creates let's say 8 http clients and keep a reference to them.  No other part of the system should be able to reinitiate this maanger.  If they want more http client, they have to create a new `httpManager`.
2. This manager should not expose the client out otherwise other parts of the system could accidentally close and dispose client without notifying the manager. 
3. This manager should handle allocating http client to make requests - i.e. all the `post`, `put`, `get` method should be on the manager, not on the actual http client.    

As you can see, this is a pretty major change.  So I would just suggest you create an array of http clients yourself in this single method, and pass the http client out to `downloadIndividualFile` method.